### PR TITLE
Logo implementation

### PR DESCRIPTION
**Pull request by Screwdgeh**
Sorry for the late PR on this! Life as a dad is pretty hectic right now...

In any case, please have a look at the implementation currently. 

In short:
- Logo implementation pretty much as previously implemented in ha-samsungtv-tizen
- A config option to configure the logo colour, by default it assumes color logo with white background (best looking one in any UI)
- No choice given while adding a new entry so far...

### CODE REVIEW

**Comment by ollo69** [comment]
Good work @Screwdgeh , I will take a look to your PR this w.e.

**Comment by jaruba** [comment]
These changes look ok to me on a purely code review level (but again, Python is far from my primary language)

I didn't get to testing them yet though. Ultimately, I think ollo should give the final approval and merge, as this is a large PR and it's mostly his code after the recent refactor.

File: custom_components/samsungtv_tizen/logo.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+                            LOGO_BASE_URL + "logo_paths.json"
+                        ) as response:
+                            url_date = datetime.strptime(
+                                response.headers.get("Last-Modified"),
+                                "%a, %d %b %Y %X %Z",
+                            ).astimezone()
+                            if url_date > file_date:
+                                self._last_check = datetime.now().astimezone()
+                                await self._download_latest_path_file()
+                    except:
```
Comment (ollo69) [inline custom_components/samsungtv_tizen/logo.py:103]: It is possible to use a more specific exception?

File: custom_components/samsungtv_tizen/logo.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    async def _download_latest_path_file(self):
+        try:
+            async with self._session.get(LOGO_BASE_URL + "logo_paths.json") as response:
+                response = (await response.read()).decode("utf-8")
+                if response:
+                    async with aiofiles.open(
+                        self._logo_file_download_path, mode="w+", encoding="utf-8"
+                    ) as paths_file:
+                        await paths_file.write(response)
+        except:
```
Comment (ollo69) [inline custom_components/samsungtv_tizen/logo.py:120]: It is possible to use a more specific exception?

File: custom_components/samsungtv_tizen/logo.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+            _LOGGER.warning(
+                "Not able to download paths file for logos. Using out-of-date paths file."
+            )
+
+    async def async_find_match(self, media_title):
+        """ Finds a match in the logo_paths file for a given media_title """
+        if media_title is not None and self._media_image_base_url is not None:
```
Comment (ollo69) [inline custom_components/samsungtv_tizen/logo.py:127]: Same comment as for row 83

File: custom_components/samsungtv_tizen/const.py [hunk 4]
```diff
@@ -69,12 +70,40 @@ class PowerOnMethod(Enum):
...
-    "3201601007250": "QizQxC7CUf.PlayMovies",   # Google Play
-    "3201606009684": "rJeHak5zRg.Spotify",      # Spotify
+    "3201512006785": "org.tizen.ignition",  # Prime Video
+    "3201901017640": "MCmYXNxgcu.DisneyPlus",  # Disney+
+    "11091000000": "4ovn894vo9.Facebook",  # Facebook
+    "3201601007250": "QizQxC7CUf.PlayMovies",  # Google Play
+    "3201606009684": "rJeHak5zRg.Spotify",  # Spotify
 }
+
+# Logo feature constants
```
Comment (ollo69) [inline custom_components/samsungtv_tizen/const.py:83]: I wonder if make sense define this constant here or if is better to leave them in the dedicated class `logo.py` and directly import from there.

Comment (Screwdgeh) [reply custom_components/samsungtv_tizen/const.py]: Moved the relevant constants all to logo.py. I kept CONF_LOGO_OPTION as part of all the CONF_ variables in const.py. I think it makes sense to keep that one there as it ties in with all the other core configuration options of the component.

Honestly, I'm fine either way... I can think of valid arguments for both approaches.

File: custom_components/samsungtv_tizen/logo.py (hunk 3 above)
Comment (Screwdgeh) [reply custom_components/samsungtv_tizen/logo.py]: True, done! Slightly more readable that way.

File: custom_components/samsungtv_tizen/logo.py (hunk 2 above)
Comment (Screwdgeh) [reply custom_components/samsungtv_tizen/logo.py]: Done! Made a distinction in errors by not being able to connect to download and writing the file to disk.

File: custom_components/samsungtv_tizen/logo.py (hunk 1 above)
Comment (Screwdgeh) [reply custom_components/samsungtv_tizen/logo.py]: Done!

File: custom_components/samsungtv_tizen/logo.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    async def _async_ensure_latest_path_file(self):
+        """ Does check if logo paths file exists and if it does - is it out of date or not. """
+        if self._media_image_base_url is not None and (
+            self._last_check is None
+            or self._last_check
+            < (
+                datetime.now().astimezone()
+                - timedelta(days=LOGO_FILE_DAYS_BEFORE_UPDATE)
+            )
+        ):
```
[the remark this answers is not in the document]
Comment (Screwdgeh) [reply custom_components/samsungtv_tizen/logo.py]: True, done!

File: custom_components/samsungtv_tizen/logo.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+            media_title = media_title.lower().strip()
+            try:
+                if os.path.isfile(self._logo_file_download_path):
+                    async with aiofiles.open(self._logo_file_download_path, "r") as f:
+                        image_paths = iter(json.loads(await f.read()).items())
+                elif os.path.isfile(self._logo_file_path):
+                    async with aiofiles.open(self._logo_file_path, "r") as f:
+                        image_paths = iter(json.loads(await f.read()).items())
+                best_match = {"ratio": None, "title": None, "path": None}
+                while True:
```
[the remark this answers is not in the document]
Comment (Screwdgeh) [reply custom_components/samsungtv_tizen/logo.py]: Exchanged this for a normal for loop and a max amount of items to evaluate before breaking (and warning if we do).

File: custom_components/samsungtv_tizen/media_player.py [hunk 7]
```diff
@@ -806,28 +818,70 @@ def icon(self):
...
                     # the channel name holds the running app ID
                     # regardless of the self._cloud_source value
-                    return self._st.channel_name
+                    new_media_title = self._st.channel_name
+                else:
+                    new_media_title = self._get_source()
+            else:
+                new_media_title = self._get_source()
+        else:
+            new_media_title = self._get_source()
```
[the remark this answers is not in the document]
Comment (Screwdgeh) [reply custom_components/samsungtv_tizen/media_player.py]: Done, new function added that uses the if constructs in combination with return statements to get the new media title; this is indeed a lot more readable.

**Comment by ollo69** [comment]
It looks good to me👍

**Review by ollo69 — APPROVED** [review]

**Comment by jaruba** [comment]
@ollo69 @Screwdgeh anything else that needs to be done in order to release? or do we just need the docs?

**Comment by ollo69** [comment]
There is just 1 think that I have in mind, that is adding `version` key in `manifest.json` This is required in last HA version.
The most complex thing is docs update, please let me know how I can help. Obviously you can take any docs from my fork that can help.

**Comment by jaruba** [comment]
Wasted a few hours today, I attempted to go through the steps of installing this component again and had no luck with it. Both my Series 6 and Series 7 TVs seem to have went in a Samsung Cloud blackhole.. They show as offline in the api.smartthings.com, the SmartThings App seems to also be unable to control either of them, and the web config of the component gets stuck at "SmartThings TV deviceID not found." even after entering it manually..

I whitelisted everything related to samsung in both Pi Hole and Ad Guard, but it still seems dead.. This is not much of a surprise though, the SmartThings API has never been very reliable and failed in random ways for many users before these changes too, but it can sure be strange to get stuck in a component setup process.. I'll try again in the next few days to see if I can make progress with it.. As I can't write the docs for it unless I'm able to install it again..

**Pull request merged**